### DISCUSSION TITLE
Eager loading joining problem when using polymorphic_identity

### DISCUSSION THREAD

**Discussion by ddorian** [sqlalchemy/sqlalchemy#8737] [Usage Questions] [answered]
I have these relationships:

```python

class Doc(db.Model):
    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    content = db.relationship(
        "Content", back_populates="doc", order_by="asc(Content.order)"
    )
```
```python
class Content(db.Model):
    __tablename__ = "content"

    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    doc_id = db.Column(UUID(as_uuid=True), db.ForeignKey("doc.id"), nullable=False)
    content_type = db.Column(db.Enum(ContentType), nullable=False)
    doc = db.relationship("Doc", back_populates="content")
    __mapper_args__ = {"polymorphic_on": content_type}

class Chapter(Content):
    __mapper_args__ = {"polymorphic_identity": "chapter"}

class OpenQuestions(Content):
    questions = db.relationship(
        "OpenQuestion",
        back_populates="content",
        order_by="asc(OpenQuestion.order)",
        lazy="raise",
    )
    __mapper_args__ = {"polymorphic_identity": "open_questions"}

```
```python
class OpenQuestion(db.Model):
    __tablename__ = "open_question"

    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    content_id = db.Column(UUID(as_uuid=True), nullable=False)
    content_type = db.Column(db.Enum(ContentType), nullable=False)

    content = db.relationship(
        "OpenQuestions",
        back_populates="questions",
    )

```

And I'm trying to make an efficient query to get all `OpenQuestions.questions` but I can't. Currently, I have:
```python
query = Doc.query
query = query.options(
    joinedload(Doc.content)
    .options(
            joinedload(OpenQuestions.questions),
        )
    )
```
But I get n+1 query of `open_question` for every row in the form of:
```
select ....
FROM open_question 
WHERE %(param_1)s = open_question.content_id AND %(param_2)s = open_question.content_type ORDER BY open_question."order" ASC
```

And using `selectinload(OpenQuestions.questions)` instead raises `lazy=raise`. 
How can I load `OpenQuestions.questions` efficiently?

regards

**Comment by zzzeek** [comment, marked as answer]
you need to either load OpenQuestions more directly in your joinedload() or use a polymorphic loading form.

The first form would be:

```py

from sqlalchemy.orm import with_polymorphic

wp = with_polymorphic(Content, [OpenQuestions])
query = Doc.query
query = query.options(
    joinedload(Doc.content.of_type(wp)).joinedoad(wp.OpenQuestions.questions)
```

or it could be against the subtype directly:


```py

query = Doc.query
query = query.options(
    joinedload(Doc.content.of_type(OpenQuestions)).joinedoad(OpenQuestions.questions)
```


another way to do it, which should emit just one extra query, using [selectin_polymorphic](https://docs.sqlalchemy.org/en/20/orm/queryguide/inheritance.html#using-selectin-polymorphic)  (i'm pointing to the 2.0 docs because I recently rewrote them, the 1.4 docs I can't even follow very well)


```py

query = Doc.query
query = query.options(
    joinedload(Doc.content).selectin_polymorphic(Content, [OpenQuestions]).joinedoad(OpenQuestions.questions))
```

all three of these, I haven't tried in a long while in this specific way, so if they aren't working right (like the last one), let me know and I'll try to figure out why.

**Reply by ddorian** [reply]
First one worked, thank you!

**Reply by zzzeek** [reply]
yeah just tried them, that last form *should* work, but it does not currently, 1.4 ignores it and 2.0 raises a more informative error. it's attempting to combine the form at https://docs.sqlalchemy.org/en/20/orm/queryguide/inheritance.html#combining-additional-loader-options-with-selectin-polymorphic-subclass-loads with the one at https://docs.sqlalchemy.org/en/20/orm/queryguide/inheritance.html#applying-selectin-polymorphic-to-an-existing-eager-load but putting them together is being rejected.  

anyway, the most direct query is like this:

```py
s = Session()
query = s.query(Doc)
query = query.options(
    joinedload(Doc.content.of_type(OpenQuestions), innerjoin=True).joinedload(
        OpenQuestions.questions, innerjoin=True
    )
)

```


the innerjoin is assuming both relationships expect to have a value present.
